### ISSUE TITLE
Debugging ng test doesn't work

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode-recipes#9]
_From @jrieken on June 27, 2017 8:43_

testing #29476

The last step which is about debugging test doesn't work for me. Breakpoints appear as empty circles not discs. 

<img width="556" alt="screen shot 2017-06-26 at 16 29 14" src="https://user-images.githubusercontent.com/1794099/27578805-7d274890-5b25-11e7-9a81-00633b777bf0.png">

_Copied from original issue: Microsoft/vscode#29539_

**Comment by weinand** [comment]
_From @chrmarti on June 27, 2017 20:45_

Same here on Windows. `ng test` spawns a Chrome window which is not mentioned in the recipe and the launch config the spawns another window that doesn't trigger the breakpoints.

**Comment by weinand** [comment]
The test setup (`ng test`) seems to have changed in a bigger way: they are now based on karma (which I think they weren't before).

/cc @roblourens @tonysneed

**Comment by roblourens** [comment]
Was looking at this with @lixire, and I couldn't even get sourcemaps loaded in chrome devtools.

**Comment by auchenberg** [comment]
Maybe related, http://blog.mlewandowski.com/Debugging-Karma-tests-with-VSCode.html?

**Comment by tonysneed** [comment]
@weinand I noticed recently that the source maps are an issue again -- this time with `ng test`.

**Comment by weinand** [comment]
@tonysneed I don't think it's just a source map issue. `ng test` now spawns a Chrome window which is not mentioned in the recipe. So the whole approach to these tests has changed.

**Comment by tonysneed** [comment]
@weinand I don't think the approach has changed in a way that can't be documented in the recipe. ng test still uses Karma like before, and the window that pops up has a debug button on it, which brings up the same url as specified in the test confit in launch.json.

So if the source map issue is resolved, we should be back on track with debugging tests with the chrome debugger extension for vs code. Can we get some input from someone on the A-CLI and/or the debugger extension teams?

**Comment by roblourens** [comment]
I can look into it later but I really don't have time this week.

**Comment by auchenberg** [comment]
Running `.scripts` with a generated angular-cli 1.1.3 app gives me this

```

› VM45
› VM47
› VM49
› VM53
› VM55
› VM57
› VM59
› VM61
› VM63
› http://localhost:9876/_karma_webpack_/inline.bundle.js
› http://localhost:9876/_karma_webpack_/main.bundle.js
› http://localhost:9876/_karma_webpack_/polyfills.bundle.js
› http://localhost:9876/_karma_webpack_/vendor.bundle.js
› http://localhost:9876/base/node_modules/jasmine-core/lib/jasmine-core/jasmine.js (/Users/auchenberg/development/angular-demo/node_modules/jasmine-core/lib/jasmine-core/jasmine.js)
› http://localhost:9876/base/node_modules/karma-jasmine-html-reporter/src/lib/adapter.js (/Users/auchenberg/development/angular-demo/node_modules/karma-jasmine-html-reporter/src/lib/adapter.js)
› http://localhost:9876/base/node_modules/karma-jasmine-html-reporter/src/lib/html.jasmine.reporter.js (/Users/auchenberg/development/angular-demo/node_modules/karma-jasmine-html-reporter/src/lib/html.jasmine.reporter.js)
› http://localhost:9876/base/node_modules/karma-jasmine-html-reporter/src/lib/html.jasmine.reporter.js (/Users/auchenberg/development/angular-demo/node_modules/karma-jasmine-html-reporter/src/lib/html.jasmine.reporter.js)
› http://localhost:9876/base/node_modules/karma-jasmine/lib/adapter.js (/Users/auchenberg/development/angular-demo/node_modules/karma-jasmine/lib/adapter.js)
› http://localhost:9876/base/node_modules/karma-jasmine/lib/boot.js (/Users/auchenberg/development/angular-demo/node_modules/karma-jasmine/lib/boot.js)
› http://localhost:9876/context.js
› http://localhost:9876/debug.html
› http://localhost:9876/debug.js
› ng:///DynamicTestModule/AppComponent.ngfactory.js
    - ng:///DynamicTestModule/AppComponent.ngfactory.js (/Users/auchenberg/development/angular-demo/DynamicTestModule/ng:/DynamicTestModule/AppComponent.ngfactory.js)
    - ng:///DynamicTestModule/AppComponent.html (/Users/auchenberg/development/angular-demo/DynamicTestModule/ng:/DynamicTestModule/AppComponent.html)
› ng:///DynamicTestModule/AppComponent_Host.ngfactory.js
    - ng:///DynamicTestModule/AppComponent_Host.ngfactory.js (/Users/auchenberg/development/angular-demo/DynamicTestModule/ng:/DynamicTestModule/AppComponent_Host.ngfactory.js)
    - ng:///DynamicTestModule/AppComponent_Host.html (/Users/auchenberg/development/angular-demo/DynamicTestModule/ng:/DynamicTestModule/AppComponent_Host.html)
› ng:///DynamicTestModule/module.ngfactory.js
    - ng:///DynamicTestModule/module.ngfactory.js (/Users/auchenberg/development/angular-demo/DynamicTestModule/ng:/DynamicTestModule/module.ngfactory.js)
```

Looks like there's some magic bundeling going on now in `ng:///DynamicTestModule/AppComponent.ngfactory.js`, as the file names are vastly different then source.

**Comment by auchenberg** [comment]
Downgrading to angular-cli 1.1.1 resolves the issue. Seems to validate this reported issue in `angular-cli`. https://github.com/angular/angular-cli/issues/6824

**Comment by roblourens** [comment]
That issue sounds very relevant :D

**Comment by weinand** [comment]
@auchenberg thanks for investigating this and finding angular/angular-cli#6824. 
I will update the recipe to make users aware of that problem for angular-cli versions > 1.1.1.
We will try to fix this in July.

**Comment by daBishMan** [comment]
first off thank you so much for your work, and your time.

Can you please add steps to debug e2e tests, using protractor and the CLI

**Comment by tonysneed** [comment]
@weinand I'm working on a PR for the recipe for setting the exact version of Angular CLI to 1.1.1.

**Comment by weinand** [comment]
@tonysneed thanks a lot

**Comment by daBishMan** [comment]
@tonysneed @weinand just for learning purpose, whey running npm start vs. ng start ?  if you have the same version of the CLI locally and globally it should be OK to go with ng start ?

**Comment by tonysneed** [comment]
@daBishMan So the issue arises when you update the global CLI package.  If you run `ng serve` then you are using the global package, which would then be a greater version that the local package.  If there is an issue with the newer version of the global package, such as broken source maps, then you may not be able to debug.

So it's much better to use `npm start` because you are guaranteed to use the local version which is set exactly (no caret or tilde) in package.json.

Also you can pass args to the npm command.  For example to open the browser:

```
npm start -- -o
```

**Comment by daBishMan** [comment]
@tonysneed thank you :)

**Comment by tonysneed** [comment]
@weinand @daBishMan Happy to help! 😄

**Comment by daBishMan** [comment]
@tonysneed thank you so much for your help with this.  

When I try to debug e2e tests I get the error below.  Yes, the breakpoint gets hit in VS Code, but the test fails.  Anyone else experiencing this?

1) biz-app App should display welcome message
  - Failed: Error while waiting for Protractor to sync with the page: "window.angular is undefined.  This could be either because this is a non-angular page or because your test involves client-side navigation, which can interfere with Protractor's bootstrapping.  See http://git.io/v4gXM for details"
Executed 1 of 1 spec (1 FAILED) in 9 secs.

**Comment by auchenberg** [comment]
@daBishMan I just clone your repo and ran `Launch ng e2e` from `VS Code 1.14.1` and things work as expected after I first ran `ng e2e` from the Terminal (installed webdriver dependencies)

```
@angular/cli: 1.2.3
node: 7.9.0
os: darwin x64
@angular/animations: 4.3.1
@angular/common: 4.3.1
@angular/compiler: 4.3.1
@angular/core: 4.3.1
@angular/forms: 4.3.1
@angular/http: 4.3.1
@angular/platform-browser: 4.3.1
@angular/platform-browser-dynamic: 4.3.1
@angular/router: 4.3.1
@angular/cli: 1.2.3
@angular/compiler-cli: 4.3.1
@angular/language-service: 4.3.1
```

<img width="1078" alt="screen shot 2017-07-24 at 3 43 30 pm" src="https://user-images.githubusercontent.com/173559/28548134-25cae63e-7087-11e7-811f-caa7e25bb5ac.png">

**Comment by Looted** [comment]
I have a related issue so I'll just ask here instead of creating another issue. The debugging works for me in such way, that when I'm stepping forward, nothing happens in the browser. Only after I step out of the it blocks, the commands actually run in the browser. Is this the desired behavior? When running e2e tests in Ruby I could step forward and see what the browser is actually doing each step (so I could inspect the DOM state etc.)
